include/sparrow/layout/list_layout/list_array.hpp

```diff
@@ -96,78 +148,208 @@ namespace sparrow
+    template <class DERIVED>
+    auto list_array_crtp_base<DERIVED>::value(size_type i) -> inner_reference
     {
-        return  list_value{p_flat_array.get(), p_list_offsets[i], p_list_offsets[i + 1]};
+        const auto r = this->derived_cast().offset_range(i);
+        using st = typename list_value::size_type;
+        return list_value{p_flat_array.get(), static_cast<st>(r.first), static_cast<st>(r.second)};
     }
 
-    template <bool BIG>
-    auto list_array_impl<BIG>::value(size_type i) const -> inner_const_reference
+    template <class DERIVED>
+    auto list_array_crtp_base<DERIVED>::value(size_type i) const -> inner_const_reference
     {
-        return list_value{p_flat_array.get(), p_list_offsets[i], p_list_offsets[i + 1]};
+        const auto r = this->derived_cast().offset_range(i);
+        using st = typename list_value::size_type;
+        return list_value{p_flat_array.get(), static_cast<st>(r.first), static_cast<st>(r.second)};
+    }
+
+    #ifdef __GNUC__
+    #pragma GCC diagnostic push
+    #pragma GCC diagnostic ignored "-Wcast-align"
+    #endif
+
+
+    template<bool BIG>
+    inline list_array_impl<BIG>::list_array_impl(arrow_proxy proxy)
+    :   base_type(std::move(proxy)),
+        p_list_offsets(reinterpret_cast<offset_type*>(this->storage().buffers()[OFFSET_BUFFER_INDEX].data() + this->storage().offset()))
+    {
+    }
+    #ifdef __GNUC__
+    #pragma GCC diagnostic pop
+    #endif
+    
+
+    template<bool BIG>
+    auto list_array_impl<BIG>::offset_range(size_type i) const -> std::pair<offset_type,offset_type>{
+        return std::make_pair(p_list_offsets[i], p_list_offsets[i+1]);
+    }
+
+    #ifdef __GNUC__
+    #pragma GCC diagnostic push
+    #pragma GCC diagnostic ignored "-Wcast-align"
+    #endif
+
+    template<bool BIG>
+    inline list_view_array_impl<BIG>::list_view_array_impl(arrow_proxy proxy)
+    :   base_type(std::move(proxy)),
+        p_list_offsets(reinterpret_cast<offset_type*>(this->storage().buffers()[OFFSET_BUFFER_INDEX].data() + this->storage().offset())),
+        p_list_sizes(reinterpret_cast<offset_type*>(this->storage().buffers()[SIZES_BUFFER_INDEX].data() + this->storage().offset()))
+    {
+    }
+
+    #ifdef __GNUC__
+    #pragma GCC diagnostic pop
+    #endif
+
+    template<bool BIG>
+    inline auto list_view_array_impl<BIG>::offset_range(size_type i) const -> std::pair<offset_type,offset_type> {
+        const auto offset = p_list_offsets[i];
+        return std::make_pair(offset, offset + p_list_sizes[i]);
+    }
+    
+    inline fixed_sized_list_array::fixed_sized_list_array(arrow_proxy proxy)
+    :   base_type(std::move(proxy)),
+        m_list_size(0)
+    {
+        // get the list size from the format string
+        const auto format = std::string(this->storage().format());
+        const auto n_digits = format.size() - 3;
+        const auto list_size_str = format.substr(3, n_digits);
+        m_list_size = std::stoull(list_size_str);
```

Comment:
Should be extracted in a function which will be used in l.342.
Also it seems that some checks are missing like format.size() >= 3

Reply:
with an assert in debug more or a real runtime check?

Reply:
I would say an assert. But I'm thinking if it should be checked in the proxy. As a constructed proxy should be a valid arrow array + schema.